### PR TITLE
struct as mapping key

### PR DESCRIPTION
**Pull request by StarkFishinator**

### CODE REVIEW

File: listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#[contract]
+mod StructAsMappingKey {
+    use starknet::get_caller_address;
+    use starknet::ContractAddress;
+    use poseidon::poseidon_hash_span;
```
Comment (StarkFishinator) [inline listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo:5]: Is this the correct function that I should use? @enitrat

File: listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+    struct User {
+        name : felt252,
+        age : felt252,
+        address : ContractAddress,
+    }
+
+    #[external]
+    fn set(key: User, value: felt252) {
+        _map::write(poseidon_hash_span(key.span()), value);
```
Comment (StarkFishinator) [inline listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo:19]: It does not work as it expects a span, but span is only for arrays and key is not an array (it's a struct), so I guess we have to use another function

File: listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo (hunk 1 above)
Comment (enitrat) [reply listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo:5]: yes

File: listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo (hunk 2 above)
Comment (enitrat) [reply listings/ch00-introduction/struct_as_mapping_key/src/lib.cairo:19]: your struct needs to derive `Serde` and then you can call `Serde::<User>::serialize(key,serialized_struct)` to make it an array (check corelib to see the correct syntax to use `Serde`

**Comment by enitrat** [comment]
Awesome! It works when deriving the `Drop` trait. Could you create a file with tests for the CI, as it's done in other examples?
Also, for the markdown file with explanations, you can structure it this way:
- Introduction (present a use case that validates this requirement - can you think of one in a context where it makes sense to hash fields together and use it as a key?)
- Motivation (why are we doing this? e.g. we can't use structs as mapping keys)
- Solution (serialize fields, hash them together, and use it as key)
- Example

This would be the perfect recipe for an easy-to-understand and meaningful example!

**Review by enitrat — APPROVED** [review]

**Pull request merged**